File: print_hanoi/print_hanoi.fs
// fsharp version

let draw_hanoi col1 col2 col3 spacing =
  let spacingstr = String.replicate spacing " "
  let dsize = 2*d + 1
  let colsz = dsize (Array.max [| Array.max col1 ; Array.max col2 ; Array.max col3 |])
  let height = col1.Length + col2.Length + col3.Length
  let draw1 idx (arr:int[]) =
     let offset = height - arr.Length
     let cindx = idx - offset
     let (dchr,dsz) = if cindx >= 0 then ("-",dsize arr.[cindx]) else ("|",1)
     let spaces = String.replicate ((colsz - dsz)/2) " "
     let disk = String.replicate dsz dchr
     printf "%s%s%s%s" spaces disk spaces spacingstr
  let rec each_disk cur = 
     if cur = height
     then ()
     else
        draw1 cur col1 ;  draw1 cur col2 ;  draw1 cur col2 ;  printfn ""
        each_disk (cur+1)
  each_disk 0

[<EntryPoint>]
let main argv =
  draw_hanoi [| 4; 5; 6; 7; |]  [| 2; 3 |]  [| 1 |]  1
  0

File: print_hanoi/print_hanoi.scala
object PrintHanoi {

  def draw( d1: Array[Int], 
            d2: Array[Int], 
            d3: Array[Int], 
            spacing: Int ): Unit = {

    val height = d1.length + d2.length + d3.length
    def formatted_size( d: Int ) = d*2+1
    val colsize = formatted_size(d1.max max d2.max max d3.max) 
    val spacingstr = " "*spacing

    def drawOne( column: Array[Int], row: Int ) = { 
       val diskidx = row - (height - column.length)
       val dsz = if ( diskidx >= 0 ) formatted_size( column(diskidx) ) else 1
       val spaces = " "*((colsize - dsz)/2)
       var chars = (if(dsz == 1) "|"  else "-")*dsz
       print(s"$spaces$chars$spaces$spacingstr")
    }

    for ( row <- 0 until height ) {
       drawOne( d1, row ) ; drawOne( d2, row ) ; drawOne( d3, row )
       println
    } 
  }  

  def main( argv: Array[String] ) = 
     draw( Array(4,5,6,7), Array(2,3), Array(1), 1)
}
